### PR TITLE
Integrate testWidgets with leak tracking.

### PR DESCRIPTION
**Pull request by polina-c**
Contributes to: https://github.com/flutter/flutter/issues/135856

TODO: 
- [x] Publish new leak tracker API
- [x] leak_tracker_flutter_testing should be refreshed in G3 (cl/590382409)
- [x] [Upgrade flutter to new leak_tracker_flutter_testing](https://github.com/flutter/flutter/pull/140018)
- [x] G3fix that adds leak_tracker_flutter_testing as dependency of testWidgets, should be added to this PR (cl/580728987)
- [x] Fix failing tests

### CODE REVIEW

File: packages/flutter_test/pubspec.yaml
Comment (polina-c) [inline packages/flutter_test/pubspec.yaml:57]: Remove this before merge.

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (goderbauer) [reply packages/flutter_test/lib/src/widget_tester.dart:205]: (see also the comment about PhaseSettings)

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (goderbauer) [reply packages/flutter_test/lib/src/widget_tester.dart:214]: Also, if this is necessary, can we add this to the existing teardown hook added in line 178?

File: packages/flutter_test/test/widget_tester_leaks_test.dart
Comment (goderbauer) [inline packages/flutter_test/test/widget_tester_leaks_test.dart:39]: formatting, here and elsewhere in this file

File: packages/flutter_test/test/widget_tester_leaks_test.dart
Comment (goderbauer) [inline packages/flutter_test/test/widget_tester_leaks_test.dart:99]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-abandonware

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (polina-c) [reply packages/flutter_test/lib/src/widget_tester.dart:13]: leak_tracker_testing contains just matcher. It is lowly coupled with the core leak_tracker.

It is separated to simplify G3 structure that wants flag 'test_only'. 
Matchers are test_only in G3, so they cannot be used for a running app. 
And, we want leak tracker to be used for running app. 

testWidgets need both leak tracker and matcher. So, it needs both packages.

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (polina-c) [reply packages/flutter_test/lib/src/widget_tester.dart:214]: Added comments to answer your questions.

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (polina-c) [reply packages/flutter_test/lib/src/widget_tester.dart:278]: I switched to tearDownAll, but there maybe not tearDownAll per test file, if tests are by groups.
So, I detect the very last tearDownAll, by counting tests.
Before we reported memory leaks per test, but it created bad performance. 
To detect notGCed leaks we need to force garbage collection. It is time consuming and impacts total test run time significantly, if done for each test.
So, we switched to bulk processing.
Each detected leak has test name, so troubleshooting UX is not impacted.

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (polina-c) [reply packages/flutter_test/lib/src/widget_tester.dart:230]: Leak tracker is designed to work for both tests and running application. It has two configuration classes: LeakTrackingConfig (is provided in the very beginning) and PhaseSetting (can be changed in process).

LeakTesting contains settings that makes sense only for testing. 

testWidgets takes LeakTesting and provides right LeakTrackingConfig and PhaseSetting to leak_tracker.

Originally it was done by `testWidgetsWithLeakTracking` defined in the package leak_tracker_flutter_testing.
Now we are moving this to `testWidgets`.

In future we may add leak tracking feature to Dart-only `test`, by providing instance of LeakTesting as a parameter. Implementation will be little different.

I do not think it is possible to create some meaningful helpers in leak_tracker or leak_tracker_flutter_testing, that will simplify code of testWidgets and incapsulate usage of LeakTrackingConfig and PhaseSetting. It seems to be over-design.
I am open for suggestions though.

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (polina-c) [reply packages/flutter_test/lib/src/widget_tester.dart:297]: Added comment.

I do not expect people to use this member. It is needed just to test `testWidgets`. It is similar to `reportTestException` that is used in flutter_test/test/widget_tester_test.dart.

Making it setting in experimentalLeakTesting will add significant complexity to the code of testWidgets, because the reporter will need to change test by test, after tests are completed, requiring to store the handlers in a global variable.

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (polina-c) [reply packages/flutter_test/lib/src/widget_tester.dart:205]: explained in other comment

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (goderbauer) [reply packages/flutter_test/lib/src/widget_tester.dart:239]: My high level feedback here is to design a proper API for leak tracking that can be used to integrate it into any testing framework. It's a bit strange that all these leak tracking internal details leak out into flutter_test. It makes the code here even more complex and harder to understand.

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (goderbauer) [inline packages/flutter_test/lib/src/widget_tester.dart:216]: typo: tha -> that

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (goderbauer) [inline packages/flutter_test/lib/src/widget_tester.dart:221]: Is it guaranteed that this will run after all user-defined tearDownAlls? In theory, those could dispose things that we want to catch as well (similar to what you describe above for tearDown), right?

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (goderbauer) [inline packages/flutter_test/lib/src/widget_tester.dart:220]: Not sure I understand this, why can we not just add this when we find the first non-ignored test?

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (goderbauer) [reply packages/flutter_test/lib/src/widget_tester.dart]: If this is only meant for internal testing of flutter_test it should at the very least be marked `@visibleForTesting`. But then again, I would encourage you to find a way to do that without adding more stuff to the global name space because ultimately that still is public API.

**Comment by polina-c** [comment]
@goderbauer ,

After creating separate hook `tearDownForTestFile` and agreeing not to configure test settings in setUp, things became much easier, and I could move most code to leak_tracker_flutter_testing. Can you check?

The PR is marked as draft, because prerequisites are not met yet and tests do not pass.

File: packages/flutter_test/lib/src/test_compat.dart
Comment (polina-c) [inline packages/flutter_test/lib/src/test_compat.dart:268]: @natebosch, created hook as we discussed. Can you check it looks ok?

It is still little hacky. An alternative would be to make `tearDownForTestFile` a first citizen and put it to declarer, but it will make the method public that we maybe do not want, because end user does not need it.

What do you think?

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (polina-c) [reply packages/flutter_test/lib/src/widget_tester.dart]: Created another hook.

**Comment by polina-c** [comment]
> I think if we were to add a specific hook for this we'd probably want to do it somewhere around the platform definition.

@natebosch , it is important that errors raised by tear down, are handled properly as test failures. 'tearDownAll' makes it happen. Will 'somewhere around the platform definition' do this?

**Comment by natebosch** [comment]
> Will 'somewhere around the platform definition' do this?

We'd make sure that is the case. I don't have any specific proposals for how we'd implement a hook in the test runner.

**Comment by polina-c** [comment]
> > Will 'somewhere around the platform definition' do this?
> 
> We'd make sure that is the case. I don't have any specific proposals for how we'd implement a hook in the test runner.

Can you link the code that you mean under 'somewhere around the platform definition'?

**Comment by natebosch** [comment]
> Can you link the code that you mean under 'somewhere around the platform definition'?

It would most likely be a callback you could pass to `deserializeSuite` to run as a suite teardown.
https://github.com/flutter/flutter/blob/cf09a8a24f52874399dd34d900b20cd4b7a6af92/packages/flutter_tools/lib/src/test/flutter_platform.dart#L352

**Comment by natebosch** [comment]
Although, thinking about it now that wouldn't let us run the code in the same isolate as the tests - I assume that's a requirement. I was thinking we could try with https://github.com/dart-lang/test/pull/2103, but I think we'd need a different place to handle it. Perhaps we should look at writing extra behavior into the bootstrapping code, so somewhere around https://github.com/flutter/flutter/blob/cf09a8a24f52874399dd34d900b20cd4b7a6af92/packages/flutter_tools/lib/src/test/flutter_platform.dart#L227-L231

File: packages/flutter_test/lib/src/widget_tester.dart
Comment (polina-c) [inline packages/flutter_test/lib/src/widget_tester.dart:213]: ignoreAllLeaks should be replaced with method like `finalizeTest`, defined in leak_tracker, that:

1. Declare all not disposed objects as leaks (it should happen for each test, not one time at the end, because we want to avoid flakiness) 
2. Pause leak tracking

File: packages/flutter_test/test/widget_tester_leaks_test.dart
Comment (natebosch) [inline packages/flutter_test/test/widget_tester_leaks_test.dart:22]: [nit] I don't think this comment is true - we also intercept `test` and others, so I would expect any test structure to work fine.

File: packages/flutter_test/test/widget_tester_leaks_test.dart
Comment (natebosch) [inline packages/flutter_test/test/widget_tester_leaks_test.dart:87]:
```suggestion
  } on TestFailure catch (e) {
```

**Review by natebosch — APPROVED** [review]
Integration with the Test runner LGTM.

File: packages/flutter_test/test/widget_tester_leaks_test.dart
Comment (polina-c) [reply packages/flutter_test/test/widget_tester_leaks_test.dart:22]: True, but we want to test that any structure will work fine. That's why in the test group goes first and then single test, to make sure this story is covered.

File: packages/flutter_test/test/widget_tester_leaks_test.dart
Comment (goderbauer) [inline packages/flutter_test/test/widget_tester_leaks_test.dart:23]: This line with the three "ignores" in it (`ignoredLeaks`, `IgnoredLeaks`, `ignore`) reads rather cryptic. I tried to figure out from the docs what this is all supposed to mean and I am left to wonder:

- why do we need to pass in empty IgnoredLeaks sets? If somebody has globally configured to ignore a certain leak (e.g. because something deep inside the framework is known to leak and we haven't cleaned that up yet), why does this test want to re-enable those leaks? Ideally, it would only track the leaks it actively creates itself, not any other leaks that on a global level were turned off? Or am I misunderstanding this?

- Should the "ignore: false" parameter be renamed to "enabled" to reflect (at least according to the docs) that it controls whether leak tracking is enabled or not? This also helps with not overusing the term "ignore" even more.

**Review by goderbauer — COMMENTED** [review]
LGTM after questions are answered and TODOs are resolved.

This integration looks a lot better! Thanks for cleaning it up.

File: packages/flutter_test/test/widget_tester_leaks_test.dart
Comment (polina-c) [reply packages/flutter_test/test/widget_tester_leaks_test.dart:23]: In this exact test we want to test leak tracking works, not to verify leaks. So, we wipe off all settings.
But, you are right: some classes may be excluded when leaking and we do not want this test to fail.

We discussed naming a lot with Kenzie and agreed to use the same word everywhere for consistency.

**Comment by polina-c** [comment]
g3 passed: https://critique.corp.google.com/cl/592664397
refreshing branch and setting next g3 to skip

**Comment by flar** [comment]
This change is causing the flutter_test_performance task to fail due to a new unexpected line in the test output that it treats as an error.

**Pull request merged**
